### PR TITLE
More EpochResultStore refactors! :tada:

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Refactoring done while debugging #5521

## Before submitting

- [n/a] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [n/a] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [n/a] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified
 - [x] **Check that target branch and milestone match!**

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 1]
```diff
@@ -281,18 +261,23 @@ def __getitem__(self, key: str) -> Any:
...
     @property
-    def has_split_and_opt_idx(self):
-        """
-        This function informs if we are running within training batch loop
-        """
-        return self._split_idx is not None and self._opt_idx is not None
-
-    @property
-    def extra_info(self):
+    def info(self):
```
Comment (ananthsub) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:264]: should this return a dataclass? right now there isn't much structure as to what's returned in this map which could lead to breakages later if keys are renamed

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 2]
```diff
@@ -148,48 +148,32 @@ def get_epoch_log_metrics(self, *_, **__) -> List[Dict]:
...
-            extra_info = {}
-
-        # [dataloader_idx][optimizer_idx][training_step_idx] is a list
-        if len(extra_info) > 0:
-            self._internal_type = ResultStoreType.INSIDE_BATCH_TRAIN_LOOP
-            # initialize dictionary
+    def append(self, result: Result, info: Dict) -> None:
+        dataloader_idx = info["dataloader_idx"]
+        self._internal_type = info["type"]
+        opt_idx = info["opt_idx"]
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:154]: shall it rather uses `.get()` also what if any of the keys is missing?

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 3]
```diff
@@ -148,48 +148,32 @@ def get_epoch_log_metrics(self, *_, **__) -> List[Dict]:
...
-            self._latest_ref[dataloader_idx][opt_idx] = result
-
-        # [dataloader_idx] is a list
+            batch_idx = info["batch_idx"]
+            self._internals[dataloader_idx].setdefault(opt_idx, {})
+            self._internals[dataloader_idx][opt_idx].setdefault(batch_idx, [])
+            self._internals[dataloader_idx][opt_idx][batch_idx].append(result)
         else:
-            self._internal_type = ResultStoreType.OUTSIDE_BATCH_TRAIN_LOOP
+            # [dataloader_idx] is a list
```
Comment (Borda) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:168]:
```suggestion
```

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py [hunk 4]
```diff
@@ -148,48 +148,31 @@ def get_epoch_log_metrics(self, *_, **__) -> List[Dict]:
...
+            self._internals[dataloader_idx][opt_idx].setdefault(batch_idx, [])
+            self._internals[dataloader_idx][opt_idx][batch_idx].append(result)
         else:
-            self._internal_type = ResultStoreType.OUTSIDE_BATCH_TRAIN_LOOP
             self._internals.setdefault(dataloader_idx, [])
             self._internals[dataloader_idx].append(result)
 
             if dataloader_idx not in self._latest_ref:
                 self._latest_ref[dataloader_idx] = {}
                 self._latest_ref[dataloader_idx][0] = {}
```
Comment (tchaton) [inline pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:173]: Might have the same trouble there no ? `self._latest_ref[dataloader_idx][opt_idx]`

File: tests/trainer/optimization/test_multiple_optimizers.py [hunk 5]
```diff
@@ -61,3 +63,79 @@ def configure_optimizers(self):
     # test loss are properly reduced
     assert torch.abs(trainer.callback_metrics["loss_2_epoch"] - torch.FloatTensor(model.loss_2).mean()) < 1e-6
     assert torch.abs(trainer.callback_metrics["loss_1_epoch"] - torch.FloatTensor(model.loss_1).mean()) < 1e-6
+
+
+def test_multiple_optimizers(tmpdir):
```
Comment (tchaton) [inline tests/trainer/optimization/test_multiple_optimizers.py:68]: Not sure to understand those tests. Would be better to log something and make sure the EpochResultStore is properly populated.

**Review by tchaton — COMMENTED** [review]
Great refactor ! Let refactor info a proper structure as recommended by @ananthsub, not sure to fully understand your tests.

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:263]: This property is only called once and within our code. It's more like an utility so In don't see the point of going the extra mile and adding a dataclass.

But we could

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 2 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:154]: We define the keys. This is not user-facing so if they aren't there either somebody broke it or we have a bug

File: tests/trainer/optimization/test_multiple_optimizers.py (hunk 5 above)
Comment (carmocca) [reply tests/trainer/optimization/test_multiple_optimizers.py:68]: These are from another file also for multioptimizer tests. I merged the files into one.

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:173]: What do you mean?

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:173]:
```suggestion
                self._latest_ref[dataloader_idx][opt_idx] = {}
```

Comment (tchaton) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:173]: I didn't try it,  but worth to check.

File: tests/trainer/optimization/test_multiple_optimizers.py (hunk 5 above)
Comment (tchaton) [reply tests/trainer/optimization/test_multiple_optimizers.py:68]: Ok !

File: pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py (hunk 4 above)
Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/epoch_result_store.py:173]: Done

**Comment by Borda** [comment]
@carmocca mind check the conflicts so we can start review :]

**Review by tchaton — APPROVED** [review]
LGTM !

**Review by awaelchli — APPROVED** [review]
